### PR TITLE
Show Policies in WeaveGitOps OSS

### LINKED ISSUE #3703: Show Policies in WeaveGitOps OSS
As a user using Weave Policy Agent in my cluster, I need to view all policies installed in the cluster. 

#### Acceptance Criteria
* We need to add a new navigation item `Policies` with the same functionality in our Enterprise product. 
* Since Weave GitOps OSS is a single cluster view, the policies page will list a single cluster policy CRs. 
* The user should be able to click on the policy and view the policy details.
* 

**Acceptance Testing Issues:**
- [ ] https://app.zenhub.com/workspaces/team-timberwolf-62d58e5eb866100020d3aeb1/issues/gh/weaveworks/weave-gitops/3729

### PR DESCRIPTION
**Pull request by Samra10**
Closes #3703 

**What changed?**

- Adding Policies List page for OSS 
- Adding Policy details page for OSS 
- APIs have been added for list Policies & get Policy Detail

image
image

**Why was this change made?**

- As a request to Show the Policies list & Policy details in WeaveGitOps OSS

**How was this change implemented?**

- By porting the feature from WGE and making the proper fixes & enhancements 
- Method `ListObjects` couldn't work because the Policies are Clustered scope and the `ListObjects` get the resources from all namespaces, so the policies will be duplicated several times on the response based on the namespaces count.

**How did you validate the change?**
- Manual Testing
- Unit-Test

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: api/core/core.proto [hunk 1]
```diff
@@ -343,3 +358,68 @@ message IsCRDAvailableRequest {
...
+
+message PolicyTargets {
+    repeated string kinds = 1;
+    repeated PolicyTargetLabel labels = 2;
+    repeated string namespaces = 3;
+}
+
+message PolicyTargetLabel { map<string, string> values = 1; }
+
+message PolicyParamRepeatedString { repeated string value = 1; }
```
Comment (waleedhammam) [inline api/core/core.proto:425]: add new line at end of file and we can use align in vs code https://marketplace.visualstudio.com/items?itemName=Chouzz.vscode-better-align for the new fields

File: core/server/policies.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/hashicorp/go-multierror"
```
Comment (waleedhammam) [inline core/server/policies.go:13]: unused package

File: core/server/policies.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	"github.com/weaveworks/weave-gitops/core/clustersmngr"
+	pb "github.com/weaveworks/weave-gitops/pkg/api/core"
+	"github.com/weaveworks/weave-gitops/pkg/server/auth"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var errRequiredClusterName = errors.New("`clusterName` param is required")
```
Comment (waleedhammam) [inline core/server/policies.go:24]: we can ditch cluster name in OSS

File: ui/App.tsx [hunk 4]
```diff
@@ -119,6 +121,15 @@ const App = () => {
           />
           <Route path={V2Routes.UserInfo} component={UserInfo} />
 
+          <Route
+            path={V2Routes.Policies}
+            component={withSearchParams(PoliciesList)}
```
Comment (TheGostKasper) [inline ui/App.tsx:126]: **PoliciesList** doesn't require Params, right ? if so we can just set
```
 component ={PoliciesList}
```

File: ui/components/Icon.tsx [hunk 5]
```diff
@@ -20,6 +20,8 @@ import PauseIcon from "@material-ui/icons/Pause";
 import PersonIcon from "@material-ui/icons/Person";
 import PlayIcon from "@material-ui/icons/PlayArrow";
 import RemoveCircleIcon from "@material-ui/icons/RemoveCircle";
+import { CallReceived, CallMade, Remove } from "@material-ui/icons";
```
Comment (TheGostKasper) [inline ui/components/Icon.tsx:23]: You'll get  a warning for **imported multiple times**
```
import CallMade from "@material-ui/icons/CallMade";
import CallReceived from "@material-ui/icons/CallReceived";
import Remove from "@material-ui/icons/Remove";
```

File: ui/components/Icon.tsx [hunk 6]
```diff
@@ -20,6 +20,8 @@ import PauseIcon from "@material-ui/icons/Pause";
...
+import { VerifiedUser, Policy } from "@material-ui/icons";
```
Comment (TheGostKasper) [inline ui/components/Icon.tsx:24]: same goes here **imported multiple times**

File: ui/components/Policies/PolicyList/PolicyTable.tsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+interface Props {
+  policies: CustomPolicy[];
+}
+
+export const PolicyTable: FC<Props> = ({ policies }) => {
+  const { isFlagEnabled } = useFeatureFlags();
+
+  policies.forEach((policy) => {
```
Comment (TheGostKasper) [inline ui/components/Policies/PolicyList/PolicyTable.tsx:28]: [POV]: It's not recommended to use `forEach` here as it's a mutator method which in this situation you're modifying the element and that might leads to unexpected behavior so using `map` instead would be recommended

File: ui/components/Policies/PolicyList/PolicyTable.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+      key={policies?.length}
+      filters={initialFilterState}
+      rows={policies}
+      fields={[
+        {
+          label: "Policy Name",
+          value: ({ clusterName, id, name }) => (
+            <Link
+              to={formatURL(V2Routes.PolicyDetailsPage, {
+                clusterName: isFlagEnabled("WEAVE_GITOPS_FEATURE_CLUSTER")
```
Comment (TheGostKasper) [inline ui/components/Policies/PolicyList/PolicyTable.tsx:66]: you can bind `clusterName` to the parameter value in line 63, right ?

File: ui/components/Policies/PolicyList/PolicyTable.tsx [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+              {
+                label: "Cluster",
+                value: "clusterName",
+                sortValue: ({ clusterName }) => clusterName,
+              },
+            ]
+          : []),
+        {
+          label: "Age",
+          value: ({ createdAt }) => moment(createdAt).fromNow(),
```
Comment (TheGostKasper) [inline ui/components/Policies/PolicyList/PolicyTable.tsx:117]: use ``` value: ({ createdAt }) => <Timestamp time={createdAt} />``` instead as it do the same job and we won't need to install & use `moment`

File: ui/lib/types.ts [hunk 10]
```diff
@@ -37,12 +37,15 @@ export enum V2Routes {
...
   ImagePolicyDetails = "/image_policy",
-
+  Policies ="/policies",
+  PolicyDetailsPage="/policy_details",
   //Image updates subRoute
   ImageRepositories = "/image_automation/repositories",
   ImagePolicies = "/image_automation/policies",
   ImageUpdates = "/image_automation/updates",
 
+  PolicyDetails="/policies/details",
```
Comment (TheGostKasper) [inline ui/lib/types.ts:47]: Do we use both  `PolicyDetailsPage` & `PolicyDetails` ?

File: ui/pages/v2/PoliciesList.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import * as React from "react";
+import styled from "styled-components";
+import Page from "../../components/Page";
+import { PolicyTable } from "../../components/Policies/PolicyList/PolicyTable";
+import { useListListPolicies } from "../../hooks/Policies";
```
Comment (TheGostKasper) [inline ui/pages/v2/PoliciesList.tsx:5]: wdyt of changing it to `useListPolicies` ?

File: core/server/policies.go [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+	return &pb.ListPoliciesResponse{
+		Policies:      policies,
+		Total:         int32(len(policies)),
+		NextPageToken: continueToken,
+		Errors:        respErrors,
+	}, nil
+}
+
+func (cs *coreServer) GetPolicy(ctx context.Context, m *pb.GetPolicyRequest) (*pb.GetPolicyResponse, error) {
+	clustersClient, err := cs.clustersManager.GetImpersonatedClientForCluster(ctx, auth.Principal(ctx), m.ClusterName)
```
Comment (serboctor) [inline core/server/policies.go:222]: I noticed in GetObject they always use `GetImpersonatedClient` regardless of clustername value, do you know why it's done like that over there? and what's the difference here?

File: core/server/policies.go [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("error getting impersonating client: %w", err)
+	}
+
+	if m.ClusterName == "" {
```
Comment (serboctor) [inline core/server/policies.go:227]: we shouldnt require clustername in OSS

File: core/server/policies.go [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+
+	policy, err = toPolicyResponse(policyCRv2beta2, m.ClusterName, true)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.GetPolicyResponse{Policy: policy}, nil
+}
+
+func getClusterPolicyKey(clusterName, policyId string) string {
```
Comment (serboctor) [inline core/server/policies.go:246]: this is only used in collectedPolicies and I don't see collectedPolicies being used anywhere, shouldn't they both be removed?

File: core/server/policies.go [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+	default:
+		return nil, fmt.Errorf("found unsupported policy parameter type %s in policy %s", param.Type, policyID)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize parameter value %s in policy %s: %w", param.Name, policyID, err)
+	}
+	return anyValue, nil
+}
+
+func toPolicyResponse(policyCRD pacv2beta2.Policy, clusterName string, fullDetails bool) (*pb.Policy, error) {
```
Comment (serboctor) [inline core/server/policies.go:78]: how about calling it `policyToPolicyRespone`?

**Review by serboctor — CHANGES_REQUESTED** [review]

File: api/core/core.swagger.json [hunk 16]
```diff
@@ -19,7 +19,7 @@
   "paths": {
     "/v1/child_objects": {
       "post": {
-        "summary": "GetChildObjects returns the children of a given object, specified by a GroupVersionKind.\nNot all Kubernets objects have children. For example, a Deployment has a child ReplicaSet, but a Service has no child objects.",
+        "summary": "*\nGetChildObjects returns the children of a given object, specified by a GroupVersionKind.\nNot all Kubernets objects have children. For example, a Deployment has a child ReplicaSet, but a Service has no child objects.",
```
[the remark this answers is not in the document]
Comment (serboctor) [reply api/core/core.swagger.json:22]: @Samra10  And why it wasnt to the summary of your APIs?

File: core/server/policies.go [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+		if err := clustersClient.List(ctx, m.ClusterName, listV2beta2, opts...); err != nil {
+			return nil, fmt.Errorf("error while listing policies for cluster %s: %w", m.ClusterName, err)
+		}
+
+		continueToken = listV2beta2.GetContinue()
+		listsV2beta2 = map[string][]client.ObjectList{m.ClusterName: {listV2beta2}}
+	}
+
+	var policies []*pb.Policy
+	collectedPolicies := map[string]struct{}{}
```
[the remark this answers is not in the document]
Comment (Samra10) [reply core/server/policies.go:194]: It was used to fix replicated policies when multiple policy versions are served, but I forgot to remove it after removing the support for v1

File: ui/components/Icon.tsx (hunk 5 above)
Comment (AsmaaNabilBakr) [reply ui/components/Icon.tsx:23]: no warnings, but I changed them to be consistent with others.

File: ui/components/Policies/PolicyList/PolicyTable.tsx (hunk 7 above)
Comment (AsmaaNabilBakr) [reply ui/components/Policies/PolicyList/PolicyTable.tsx:28]: good point

File: ui/components/Policies/PolicyList/PolicyTable.tsx (hunk 8 above)
Comment (AsmaaNabilBakr) [reply ui/components/Policies/PolicyList/PolicyTable.tsx:66]: i update the `ClusterName` value in the same function that apply `audit` and `enforce` values

File: ui/lib/types.ts (hunk 10 above)
Comment (AsmaaNabilBakr) [reply ui/lib/types.ts:47]: no need for `PolicyDetails`  I forgot to remove it.

**Comment by jpellizzari** [comment]
@Samra10 Would it be possible to split policies into it's own `gRPC` server that is separate from `Core`? I think that would make things more modular and portable. Here is an example implementation for Terraform objects:

https://github.com/weaveworks/weave-gitops-enterprise/blob/387bb3456f1c82c500714542f755c03106b010a6/pkg/terraform/server.go#L43

WDYT?

**Comment by serboctor** [comment]
@Samra10 I had a chat with @jpellizzari regarding his refactor suggestion and we agreed to have a separate follow-up ticket for it since it will also affect other open PRs, and might block other stories for longer.

**Review by yiannistri — COMMENTED** [review]
Can we add unit tests for the 2 new endpoints being added?

**Review by TheGostKasper — APPROVED** [review]
[UI] LGTM

File: core/server/policies.go [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+func (cs *coreServer) GetPolicy(ctx context.Context, m *pb.GetPolicyRequest) (*pb.GetPolicyResponse, error) {
+	var clustersClient clustersmngr.Client
+	var err error
+
+	if m.ClusterName != "" {
+		clustersClient, err = cs.clustersManager.GetImpersonatedClientForCluster(ctx, auth.Principal(ctx), m.ClusterName)
+	} else {
+		m.ClusterName = DefaultCluster
```
Comment (serboctor) [inline core/server/policies.go:238]: why dont you just set the clusterName if it's empty and then use the same client for both cases.

**Review by serboctor — APPROVED** [review]

**Pull request merged**